File: python-monocle/Demos/pong.py
from Monocle import *
Monocle = Monocle()
class Paddle(Entity):
        def __init__(self):
                Entity.__init__(self)
                self.speed = 0.0
                self.accel = 2000.0
                self.maxSpeed = 300.0
                self.friction = 500.0
                self.maxY = 600.0
                self.keyUp = None
                self.keyDown = None
                self.AddTag("Paddle")
                rectangleCollider = RectangleCollider(25.0,100.0)
                self.SetCollider(rectangleCollider)
        def Update(self):
                Entity.Update(self)
                if(Input.IsKeyHeld(self.keyUp)):
                        self.speed += self.accel * Monocle.deltaTime
                        if(self.speed > self.maxSpeed):
                                self.speed = self.maxSpeed
                elif (Input.IsKeyHeld(self.keyDown)):
                        self.speed -= self.accel * Monocle.deltaTime
                        if(self.speed < -self.maxSpeed):
                                self.speed = -self.maxSpeed
                else:
                        if(self.speed > 0.0):
                                self.speed -= self.friction * Monocle.deltaTime
                        elif (self.speed < 0.0):
                                self.speed += self.friction * Monocle.deltaTime
                        if(abs(self.speed) < 0.05):
                                self.speed = 0.0
                self.position += Vector2.up * Monocle.deltaTime * self.speed
                if(self.position.y > self.maxY):
                        self.position.y = self.maxY
                        self.speed = -1.0
                elif(self.position.y < 0):
                        self.position.y = 0
                        self.speed = 1.0
        def Render(self):
                Entity.Render(self)
                Graphics.PushMatrix()
                Graphics.Translate(self.position)
                Graphics.RenderQuad(25.0,100.0)
                Graphics.PopMatrix();

class Ball(Entity):
        def __init__(self):
                Entity.__init__(self)
                self.texture = None
                self.AddTag("Ball")
                rectangleCollider = RectangleCollider(25.0,25.0)
                self.SetCollider(rectangleCollider)
                self.velocity = Vector2.right * 200.0
        def Update(self):
                Entity.Update(self)
                lastPosition = self.position
                self.position += self.velocity * Monocle.deltaTime
                collider = self.Collide("Paddle")
                if(collider):
                        entity = collider.GetEntity()
                        self.position = lastPosition
                        diff = self.position - collider.GetEntity().position
                        diff.Normalize()
                        diff *= self.velocity.GetMagnitude()
                        self.velocity = diff
                #if we hit the roof or floor, bounce!
                if(self.position.y < 0 or self.position.y > 600):
                        self.position = lastPosition
                        self.velocity.y *= -1
                #if we reach a player's goal, reset the ball!
                if(self.position.x < 0):
                        self.SendNoteToScene("BallOffLeft")
                if(self.position.x > 800):
                        self.SendNoteToScene("BallOffRight")
        def Render(self):
                Entity.Render(self)
                Graphics.BindTexture(self.texture)
                Graphics.PushMatrix();
                Graphics.Translate(self.position)
                Graphics.RenderQuad(25.0,25.0)
                Graphics.PopMatrix();

class GameScene(Scene):
        def __init__(self):
                Scene.__init__(self)
                self.ball = None
                self.paddle1 = None
                self.paddle2 = None
        def Begin(self):
                #Debug.Log("Pong::GameScene::Begin()!")
                print "Pong::GameScene::Begin()!"
                Scene.Begin(self)
                #do pong specific init
                self.ball = Ball()
                self.ball.position = Vector2(400,300)
                self.Add(self.ball)
                self.paddle1 = Paddle()
                self.paddle1.position = Vector2(100,300)
                self.Add(self.paddle1)
                self.paddle1.keyUp = KeyCode.KEY_W
                self.paddle1.keyDown = KeyCode.KEY_S
                self.paddle2 = Paddle()
                self.paddle2.position = Vector2(700,300)
                self.Add(self.paddle2)
                self.paddle2.keyUp = KeyCode.KEY_UP
                self.paddle2.keyDown = KeyCode.KEY_DOWN
        def ResetBall(self):
                self.ball.position = Vector2(400,300)
                randomDirection = Vector2.Random()
                #to make sure the ball always starts at the same
                #velocity, we set direction to 1 or -1
                #python has a weird ternary operator :D
                randomDirection.x = -1 if randomDirection.x < 0 else 1
                self.ball.velocity.y = 0.0
                self.ball.velocity.x = randomDirection.x * 200.0
        def ReceiveNote(self,message):
                if(message == "BallOffLeft" or message == "BallOffRight"):
                        self.ResetBall()
        def End(self):
                Scene.End(self)
                print "bye!"

game = Game()
game.Init()
scene = GameScene()
game.SetScene(scene)
game.Main()

File: Samples/Pong/Pong.cpp
#include "Pong.h"
#include <Input.h>
#include <Collision.h>

#include <stdlib.h>
#include <sstream>

namespace Pong
{    
    AudioAsset *sfxWall=NULL;

	Text::Text(const std::string& text, FontAsset* font)
		: Entity(), font(font), text(text)
	{
	}

	void Text::Render()
	{
		Graphics::PushMatrix();
		Graphics::Translate(position);
		Graphics::SetBlend(BLEND_ALPHA);
		Graphics::SetColor(Color::white);
		Graphics::BindFont(font);

		Graphics::RenderText(*font, text, 0, 0);
		Graphics::PopMatrix();
	}

    
	/*
	**************************************************
	* B a l l
	**************************************************
	*
	* Moves via velocity Vector2
	* Collides with entities tagged with "Paddle"
	* Sends a message to the scene when it goes offside
	* 
	*/

	Ball::Ball() : Entity(), texture(NULL)
	{
		AddTag("Ball");
		SetCollider(new RectangleCollider(25.0f, 25.0f));
		//Collision::AddRectangleCollider(this, 25.0f, 25.0f);
		velocity = Vector2::right * 200.0f;
	}

	void Ball::Update()
	{
		// store our last position
		Vector2 lastPosition = position;

		// update our current position using velocity
		position += velocity * Monocle::deltaTime;

		// check collisions against the paddles
		Collider* collider = Collide("Paddle");
		if (collider)
		{
			Debug::Log("Ball hit an entity tagged with 'Paddle'");
			position = lastPosition;
			
			Vector2 diff = position - collider->GetEntity()->position;
			diff.Normalize();
			diff *= velocity.GetMagnitude();
			velocity = diff;
            
            // Calculate panning
            float pan = ((collider->GetEntity()->position.x / Graphics::GetVirtualWidth()) - 0.5) * 2.0;

            if (sfxWall)
                sfxWall->Play(1,1.0,pan); // Play it with panning! (STEREO, baby :D)
		}

		// if we hit the top or bottom of the screen
		if (position.y < 0 || position.y > 600)
		{
			position = lastPosition;
			velocity.y *= -1;
            
            // Calculate panning
            float pan = ((position.x / Graphics::GetVirtualWidth()) - 0.5) * 2.0;
            
            if (sfxWall)
                sfxWall->Play(1,1.0,pan,2.0); // Play it higher, you won't even notice it's the same ;D
		}
	}

	void Ball::Render()
	{
		Graphics::BindTexture(texture);
		Graphics::PushMatrix();
		Graphics::Translate(position);
        Graphics::SetColor(Color::white);
		Graphics::RenderQuad(25.0f, 25.0f);
		Graphics::PopMatrix();
	}

	
	/*
	**************************************************
	* P a d d l e
	**************************************************
	*
	* Moves up and down, controlled by user input
	* Is tagged with "Paddle"
	* 
	*/

	Paddle::Paddle()
		: Entity(), speed(0.0f)
	{
		AddTag("Paddle");
		SetCollider(new RectangleCollider(25.0f, 100.0f));
	}

	void Paddle::Update()
	{
		const float accel = 2000.0f;
		const float maxSpeed = 300.0f;
		const float friction = 500.0f;
		const float maxY = 600.0f;

		if (Input::IsKeyHeld(keyUp) || Input::IsTouchInRect(Vector2(position.x-150,position.y-500),
                                                            Vector2(position.x+150,position.y)))
		{
			speed += accel * Monocle::deltaTime;
			if (speed > maxSpeed) speed = maxSpeed;
		}
		else if (Input::IsKeyHeld(keyDown) || Input::IsTouchInRect(Vector2(position.x-150,position.y+1),
                                                                   Vector2(position.x+150,position.y+500)))
		{
			speed -= accel * Monocle::deltaTime;
			if (speed < -maxSpeed) speed = -maxSpeed;
		}
		else
		{
			if (speed > 0.0f)
			{
				speed -= friction * Monocle::deltaTime;
			}
			else if (speed < 0.0f)
			{
				speed += friction * Monocle::deltaTime;
			}

			if (abs(speed) < 0.05f)
			{
				speed = 0.0f;
			}
		}
		position += Vector2::up * Monocle::deltaTime * speed;
		
		if (position.y > maxY)
		{
			position.y = maxY;
			speed = -1.0f;
		}
		else if (position.y < 0)
		{
			position.y = 0;
			speed = 1.0f;
		}
	}

	void Paddle::Render()
	{
		Graphics::PushMatrix();
		Graphics::Translate(position);
        Graphics::SetColor(Color::white);
		Graphics::RenderQuad(25, 100.0f);
		Graphics::PopMatrix();
	}


	/*
	**************************************************
	* G a m e S c e n e
	**************************************************
	*
	* Sets up the game's entities in Begin()
	* by creating a ball and two paddles
	* 
	* Handles messages sent to it indicating when the
	* ball has gone offside.
	* 
	*/

	void GameScene::Begin()
	{
		Debug::Log("Pong::GameScene::Begin()!");

		Scene::Begin();
        
        sfxWall = Assets::RequestAudio("AudioTest/Explosion.wav");
		
		p1Score = p2Score = 0;

		ball = new Ball();
		ball->position = Vector2(400, 300);
		Add(ball);

		paddle1 = new Paddle();
		paddle1->position = Vector2(100, 300);
		paddle1->keyUp = KEY_W;
		paddle1->keyDown = KEY_S;
		Add(paddle1);

		paddle2 = new Paddle();
		paddle2->position = Vector2(700, 300);
		paddle2->keyUp = KEY_UP;
		paddle2->keyDown = KEY_DOWN;
		Add(paddle2);

		FontAsset* font = Assets::RequestFont("Pong/LiberationSans-Regular.ttf", 25.0f);
        scoreText = new Text(GetScoreString(), font);
        scoreText->position = Vector2(50, 50);
        Add(scoreText);
	}

	void GameScene::ResetBall()
	{
		ball->position = Vector2(400,300);
		ball->velocity = Vector2::Random() * 200.0f;
	}
    
    const std::string GameScene::GetScoreString()
    {
        std::stringstream str;
        str << "P1:  " << p1Score << " - " << " P2: " << p2Score;
        return str.str();
    }

	void GameScene::Update()
	{
		Scene::Update();

		// do pong specific update
		if (ball->position.x < 0)
		{
            p2Score++;
            scoreText->SetText(GetScoreString());
			ResetBall();
		}
		else if (ball->position.x > 800)
		{
            p1Score++;
            scoreText->SetText(GetScoreString());
			ResetBall();
		}
	}

	void GameScene::End()
	{
		Scene::End();
	}
}
